cmd/cosign/cli/verify/verify_bundle.go

```diff
@@ -0,0 +1,163 @@
+//
+// Copyright 2024 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package verify
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	sgbundle "github.com/sigstore/sigstore-go/pkg/bundle"
+	"github.com/sigstore/sigstore-go/pkg/fulcio/certificate"
+	"github.com/sigstore/sigstore-go/pkg/root"
+	"github.com/sigstore/sigstore-go/pkg/verify"
+
+	"github.com/sigstore/cosign/v2/pkg/cosign/pivkey"
+	sigs "github.com/sigstore/cosign/v2/pkg/signature"
+)
+
+type verifyTrustedMaterial struct {
+	root.TrustedMaterial
+	keyTrustedMaterial root.TrustedMaterial
+}
+
+func (v *verifyTrustedMaterial) PublicKeyVerifier(hint string) (root.TimeConstrainedVerifier, error) {
+	return v.keyTrustedMaterial.PublicKeyVerifier(hint)
+}
+
+func verifyNewBundle(ctx context.Context, bundlePath, trustedRootPath, keyRef, slot, certOIDCIssuer, certOIDCIssuerRegex, certIdentity, certIdentityRegexp, githubWorkflowTrigger, githubWorkflowSHA, githubWorkflowName, githubWorkflowRepository, githubWorkflowRef, artifactRef string, sk, ignoreTlog, useSignedTimestamps, ignoreSCT bool) error {
+	if certOIDCIssuerRegex != "" {
+		return fmt.Errorf("--new-bundle-format does not support --certificate-oidc-issuer-regexp")
+	}
+
+	bundle, err := sgbundle.LoadJSONFromPath(bundlePath)
+	if err != nil {
+		return err
+	}
+
+	var trustedroot *root.TrustedRoot
+
+	if trustedRootPath == "" {
+		// Assume we're using public good instance; fetch via TUF
+		trustedroot, err = root.FetchTrustedRoot()
+		if err != nil {
+			return err
+		}
+	} else {
+		trustedroot, err = root.NewTrustedRootFromPath(trustedRootPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	trustedmaterial := &verifyTrustedMaterial{TrustedMaterial: trustedroot}
+
+	// See if we need to wrap trusted root with provided key
+	if keyRef != "" {
+		signatureVerifier, err := sigs.PublicKeyFromKeyRef(ctx, keyRef)
+		if err != nil {
+			return err
+		}
+
+		newExpiringKey := root.NewExpiringKey(signatureVerifier, time.Time{}, time.Time{})
```

Comment:
Out of scope for this, should we create a non-expiring key as well?

Reply:
The terminology here is confusing - despite the name `NewExpiringKey`, since the `validityPeriodStart` and `validityPeriodEnd` will return true for `IsZero()`, this key won't actually be treated as expiring.

In the future, if there are cosign CLI flags for specifying the validity period of the key, we could make use of them here!